modules/anPspParamsConverter.js

```diff
@@ -0,0 +1,142 @@
+/*
+- register a hook function on the makeBidRequests hook (after the main function ran)
+
+- this hook function will:
+1. verify s2sconfig is defined and we (or our aliases) are included to the config
+2. filter bidRequests that match to our bidderName or any registered aliases
+3. for each request, read the bidderRequests.bids[].params to modify the keys/values
+  a. in particular change the keywords structure, apply underscore casing for keys, adjust use_payment_rule name, and convert certain values' types
+  b. will import some functions from the anKeywords library, but ideally should be kept separate to avoid including this code when it's not needed (strict client-side setups) and avoid the rest of the appnexus adapter's need for inclusion for those strictly server-side setups.
+*/
+
+import {
+  isArray,
+  isPlainObject,
+  isStr
+} from '../src/utils.js';
+import {
+  getHook
+} from '../src/hook.js';
+import {
+  config
+} from '../src/config.js';
+import {
+  convertCamelToUnderscore,
+  appnexusAliases
+} from '../libraries/appnexusUtils/anUtils.js';
+import {
+  convertTypes
+} from '../libraries/transformParamsUtils/convertTypes.js';
+
+import adapterManager from '../src/adapterManager.js';
+
+// keywords: { 'genre': ['rock', 'pop'], 'pets': ['dog'] } goes to 'genre=rock,genre=pop,pets=dog'
+function convertKeywordsToString(keywords) {
+  let result = '';
+  Object.keys(keywords).forEach(key => {
+    // if 'text' or ''
+    if (isStr(keywords[key])) {
+      if (keywords[key] !== '') {
+        result += `${key}=${keywords[key]},`
+      } else {
+        result += `${key},`;
+      }
+    } else if (isArray(keywords[key])) {
+      if (keywords[key][0] === '') {
+        result += `${key},`
+      } else {
+        keywords[key].forEach(val => {
+          result += `${key}=${val},`
+        });
+      }
+    }
+  });
+
+  // remove last trailing comma
+  result = result.substring(0, result.length - 1);
+  return result;
+}
+
+function digForAppNexusBidder(s2sConfig) {
+  let result = false;
+  // check for plain setup
+  if (s2sConfig?.bidders?.includes('appnexus')) result = true;
+
+  // check for server-side alias
+  if (!result && isPlainObject(s2sConfig?.extPrebid?.aliases) && (Object.values(s2sConfig?.extPrebid?.aliases).includes('appnexus'))) result = true;
+
+  // registered aliases
+  const aliasList = appnexusAliases.map(aliasObj => (aliasObj.code));
+  if (!result && s2sConfig?.bidders?.filter(s2sBidder => aliasList.includes(s2sBidder)).length > 0) result = true;
+
+  // pbjs.aliasBidder
+  if (!result) {
+    result = !!(s2sConfig?.bidders?.find(bidder => (adapterManager.resolveAlias(bidder) === 'appnexus')));
+  }
+
+  return result;
+}
+
+// need a separate check b/c we're checking a specific bidRequest to see if we modify it, not just that we have a server-side bidder somewhere in prebid.js
+function isThisOurBidderInDisguise(tarBidder, s2sConfig) {
```

Comment:
This can be simplified - and I'm not sure it works in the s2sTesting scenario (where I believe a bidder can be configured for s2s, but sometimes routed to the client)

bids routed to PBS have `src: CONTANTS.S2S.SRC`. so you could do something like

```javascript
bidderRequests
  .flatMap(br => br.bids)
  .filter(bid => bid.src === CONTANTS.S2S.SRC && adapterManager.resolveAlias(bid.bidder) === 'appnexus')
  .forEach((bid) => {
      // transform bid.params
  })
```

this does not look at `s2sConfig.extPrebid`, but neither does the [original logic](https://github.com/prebid/Prebid.js/blob/78f93cd29ee0955c6ddf2cf20f5251ac8e8c66e1/libraries/pbsExtensions/processors/params.js#L9-L13). Adding that is tricky because a publisher could theoretically configure multiple s2s configs where the same alias points to appnexus only on some of them.

Reply:
Hi @dgirardi 
I updated the logic here with your suggestions.  Please take a look when you have the chance.  Thanks.